apps/passport-server/src/services/generic-issuance/pipelines/PODPipeline/utils/atomize.ts

```diff
@@ -0,0 +1,69 @@
+import {
+  InputColumn,
+  InputRow,
+  PODPipelineOutput,
+  getInputToPODValueConverter
+} from "@pcd/passport-interface";
+import { PODEntries, serializePODEntries } from "@pcd/pod";
+import { assertUnreachable } from "@pcd/util";
+import { v5 as uuidv5 } from "uuid";
+import { PODAtom } from "../PODPipeline";
+
+/**
+ * Creates a {@link PODAtom} from the output configuration, input columns, and
+ * input row. Maps input columns to POD values and handles conversions between
+ * input types and POD types.
+ *
+ * @param columns The input columns, which contain data type information
+ * @param row The input row, which contains the actual values
+ * @param output The output configuration, which contains the output entries and the match information
+ * @param outputId The ID of the output
+ * @param pipelineId The ID of the pipeline
+ * @returns The PODAtom
+ */
+export function atomize(
+  columns: Record<string, InputColumn>,
+  row: InputRow,
+  output: PODPipelineOutput,
+  outputId: string,
+  pipelineId: string
+): PODAtom {
+  const entries: PODEntries = {};
+  for (const [key, entry] of Object.entries(output.entries)) {
+    const source = entry.source;
+    if (source.type === "input") {
+      const column = columns[source.name];
+      /**
+       * See {@link getInputToPODValueConverter} for detail on how this
+       * conversion is done.
+       */
+      const converter = getInputToPODValueConverter(column.type, entry.type);
+      if (!converter) {
+        throw new Error(
+          `No converter for input ${source.name} of type ${column.type} to POD value ${key} of type ${entry.type}`
+        );
+      }
+      const cell = column.getValue(row);
+      entries[key] = converter(cell);
+    } else if (source.type === "configured") {
+      entries[key] = {
+        // @todo non-string configured values
+        type: "string",
+        value: source.value
+      };
+    } else if (
+      source.type === "credentialSemaphoreID" ||
+      source.type === "credentialEmail"
+    ) {
+      // These values are not present during loading and so cannot be
+      // populated in the Atom.
+      continue;
+    } else {
+      assertUnreachable(source);
+    }
+  }
+
+  const id = uuidv5(serializePODEntries(entries), pipelineId);
```

Comment:
Two comments to discuss here:

1) Depending on what this ID is for, I wonder if you really want a content-based ID, or if there should be another stable ID based on a specific column of the input.  E.g. if this becomes the PCD ID, it controls whether a change to one entry should cause a POD to be replaced, or treated as a whole new POD.  (This is most visible if the pipeline doesn't always delete its folder.)

2) Assuming you want a content-based ID, note that this approach doesn't canonicalize the POD entries (e.g. it will depend on input order).  If want this to be maximally stable, one option would be to use `PODContent.fromEntries(entries).contentID` which should act as a stable hash based on the Merkle tree.  This step also has the side-effect of validating that all the POD entries are legal (e.g. ints are in valid range), so you'll get an exception here if you do that (rather than getting it later at POD signing).  if you do this I'd also recommend you return the PODContent directly, or else return `content.asEntries()` which will be in canonical order.

Reply:
I'm not sure that `contentID` is a good idea here, because it requires building the merkle tree for the POD. Benchmarking a POD with 10 entries shows that it takes around 3.8ms on my laptop (which has faster single-core speed than our main server) as opposed to 0.08ms for the current approach.

This means that 1,000 PODs would take 3.8 seconds, and with 10 such pipelines 38 seconds, and we have to do this operation once per minute!

This does call into question the merits of reloading pipelines once per minute, but this is another of those cross-cutting concerns that can't be fixed for this pipeline alone and deserves an issue of its own.

Reply:
Good to know!  I didn't think of our Poseidon hashes being so much more expensive than other kinds of hashes.
Note that if you were actually building a PODContent or POD here then the Merklization would be memoized for the time later when the POD needs to be signed.  Arguably if you're reloading every minute, though, the number of reloads will usually dominate vs. the number of signatures.

Reply:
As I mentioned below, sorting the entries somewhere might be good to ensure canonicalization without incurring too much cost.  Otherwise I'd expect the value to be stable enough to be used as a content-based ID without full Merklization.

Reply:
Entries are now sorted.